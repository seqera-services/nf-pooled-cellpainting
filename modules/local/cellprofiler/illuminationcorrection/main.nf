process CELLPROFILER_ILLUMINATIONCORRECTION {
    tag "${meta.id}_${channels}"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'https://depot.galaxyproject.org/singularity/cellprofiler:4.2.8--pyhdfd78af_0'
        : 'community.wave.seqera.io/library/cellprofiler:4.2.8--aff0a99749304a7f'}"

    input:
    tuple val(meta), val(channels), path(images, stageAs: "images/*"), path(load_data_csv)

    path illumination_cppipe

    output:
    tuple val(meta), path("illumination_corrections/*.npy"), emit: illumination_corrections
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    """
    mkdir -p illumination_corrections
    # Replace the channel name in the cppipe file
    sed 's/{{channel}}/${channels}/g' ${illumination_cppipe} > illumination.cppipe

    cellprofiler -c -r \
    ${args} \
    -p illumination.cppipe \
    -o illumination_corrections \
    --data-file=${load_data_csv} \
    --image-directory ./images/ \
    -g Metadata_Plate=${meta.plate} \

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        cellprofiler: \$(cellprofiler --version |& sed '1!d ; s/cellprofiler //')
    END_VERSIONS
    """

    stub:
    """
    mkdir -p illumination_corrections
    echo 'this is not an illumination correction' > illumination_corrections/${meta.plate}_Illum${meta.channel}.npy

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        cellprofiler: \$(cellprofiler --version |& sed '1!d ; s/cellprofiler //')
    END_VERSIONS
    """
}
