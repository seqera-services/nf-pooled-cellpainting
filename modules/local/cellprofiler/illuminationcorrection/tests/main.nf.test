nextflow_process {

    name "Test Process CELLPROFILER_ILLUMINATIONCORRECTION"
    script "../main.nf"
    process "CELLPROFILER_ILLUMINATIONCORRECTION"

    tag "cellprofiler"
    tag "cellprofiler/illuminationcorrection"

    test("cellprofiler - illuminationcorrection - multi-channel - not parallel") {

        when {
            params {
                cp_multichannel_parallel = false
            }
            process {
                """
                input[0] = [
                    [batch:'Batch1', plate:'Plate1', channels:'CHN2-AF488', id:'Batch1_Plate1_CHN2-AF488'], 'CHN2-AF488',
                    [   file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA1_PointA1_0000_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq0000.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA1_PointA1_0001_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq0001.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA1_PointA1_0002_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq0002.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA1_PointA1_0003_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq0003.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA2_PointA2_0000_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq1025.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA2_PointA2_0001_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq1026.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA2_PointA2_0002_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq1027.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA2_PointA2_0003_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq1028.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellB1_PointB1_0000_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq3075.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellB1_PointB1_0001_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq3076.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellB1_PointB1_0002_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq3077.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellB1_PointB1_0003_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq3078.ome.tiff")],
                        file("${projectDir}/modules/local/cellprofiler/illuminationcorrection/tests/load_data.csv", checkIfExists: true)
                ]
                input[1] = "${projectDir}/assets/cellprofiler/illumination.cppipe.template"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.illumination_corrections).match() }
            )
        }

    }

        test("cellprofiler - illuminationcorrection - multi-channel - in parallel") {

        when {
            params {
                cp_multichannel_parallel = true
            }
            process {
                """
                input[0] = [
                    [batch:'Batch1', plate:'Plate1', channels:'CHN2-AF488', id:'Batch1_Plate1_CHN2-AF488'], 'CHN2-AF488',
                    [   file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA1_PointA1_0000_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq0000.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA1_PointA1_0001_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq0001.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA1_PointA1_0002_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq0002.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA1_PointA1_0003_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq0003.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA2_PointA2_0000_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq1025.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA2_PointA2_0001_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq1026.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA2_PointA2_0002_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq1027.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA2_PointA2_0003_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq1028.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellB1_PointB1_0000_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq3075.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellB1_PointB1_0001_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq3076.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellB1_PointB1_0002_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq3077.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellB1_PointB1_0003_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq3078.ome.tiff")],
                        file("${projectDir}/modules/local/cellprofiler/illuminationcorrection/tests/load_data.csv", checkIfExists: true)
                ]
                input[1] = "${projectDir}/assets/cellprofiler/illumination.cppipe.template"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.illumination_corrections).match() }
            )
        }

    }

    test("cellprofiler - illuminationcorrection - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = [
                    [batch:'Batch1', plate:'Plate1', channels:'CHN2-AF488', id:'Batch1_Plate1_CHN2-AF488'], 'CHN2-AF488',
                    [   file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA1_PointA1_0000_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq0000.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA1_PointA1_0001_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq0001.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA1_PointA1_0002_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq0002.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA1_PointA1_0003_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq0003.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA2_PointA2_0000_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq1025.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA2_PointA2_0001_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq1026.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA2_PointA2_0002_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq1027.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellA2_PointA2_0003_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq1028.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellB1_PointB1_0000_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq3075.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellB1_PointB1_0001_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq3076.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellB1_PointB1_0002_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq3077.ome.tiff"),
                        file("s3://nf-pooled-cellpainting-sandbox/data/test-data/fix-s1/Source1/Batch1/images/Plate1/20X_CP_Plate1_20240319_122800_179/WellB1_PointB1_0003_ChannelPhalloAF750,CHN2-AF488,DAPI_Seq3078.ome.tiff")],
                        file("${projectDir}/modules/local/cellprofiler/illuminationcorrection/tests/load_data.csv", checkIfExists: true)
                ]
                input[1] = "${projectDir}/assets/cellprofiler/illumination.cppipe.template"
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
