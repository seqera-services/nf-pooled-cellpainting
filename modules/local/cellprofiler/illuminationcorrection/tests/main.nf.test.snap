{
    "cellprofiler - illuminationcorrection - multi-channel - not parallel": {
        "content": [
            [
                [
                    {
                        "batch": "Batch1",
                        "plate": "Plate1",
                        "channels": "CHN2-AF488",
                        "id": "Batch1_Plate1_CHN2-AF488"
                    },
                    "Plate1_IllumCHN2-AF488.npy:md5,f0b7db042aba9fe16e005d38df1460a4"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-25T12:18:14.638559"
    },
    "cellprofiler - illuminationcorrection - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "batch": "Batch1",
                            "plate": "Plate1",
                            "channels": "CHN2-AF488",
                            "id": "Batch1_Plate1_CHN2-AF488"
                        },
                        "Plate1_Illumnull.npy:md5,1950d1deb8bc1c4cc1c859cd0174ecc3"
                    ]
                ],
                "1": [
                    "versions.yml:md5,36342e2e6a8f1a1ef3b02615083687aa"
                ],
                "illumination_corrections": [
                    [
                        {
                            "batch": "Batch1",
                            "plate": "Plate1",
                            "channels": "CHN2-AF488",
                            "id": "Batch1_Plate1_CHN2-AF488"
                        },
                        "Plate1_Illumnull.npy:md5,1950d1deb8bc1c4cc1c859cd0174ecc3"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,36342e2e6a8f1a1ef3b02615083687aa"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-25T12:19:56.222456"
    },
    "cellprofiler - illuminationcorrection - multi-channel - in parallel": {
        "content": [
            [
                [
                    {
                        "batch": "Batch1",
                        "plate": "Plate1",
                        "channels": "CHN2-AF488",
                        "id": "Batch1_Plate1_CHN2-AF488"
                    },
                    "Plate1_IllumCHN2-AF488.npy:md5,f0b7db042aba9fe16e005d38df1460a4"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-08-25T12:22:18.782499"
    }
}